src/kvstore/wal/FileBasedWal.cpp

```diff
@@ -238,7 +232,28 @@ void FileBasedWal::scanAllWalFiles() {
 
         // We now get all necessary info
         close(fd);
-        walFiles_.insert(std::make_pair(startIdFromName, info));
+    }
+
+    if (!walFiles_.empty()) {
+        auto it = walFiles_.rbegin();
+        if (it->second->lastId() == 0) {
+            // if last wal is empty, scan previous wal
+            auto expectedLastWalId = it->second->firstId() - 1;
+            unlink(it->second->path());
+            walFiles_.erase(it->first);
+            it = walFiles_.rbegin();
+            scanLastWal(it->second, it->second->firstId(), expectedLastWalId);
+        } else {
+            // scan last wal, if it is illegal, scan the previous wal
+            scanLastWal(it->second, it->second->firstId(), -1);
```

Comment:
Too verbose. 
Why you need the expectedLastWalId?

Reply:
That's because we didn't truncate when we rollback, and it makes verbose. Say we have a wal `1000.wal` with `firstId = 1000, lastId =1200`, and we rollback to 1100, it still has all these wal. `expectedLastWalId` is used to mark the last valid wal. 

I'will try to simplify the logic.